src/ddeinterp.ml

```diff
@@ -5,183 +5,256 @@ exception Unreachable
+  | IntResult i -> i
+  | OpResult op_r -> (
+      match op_r with
+      | Plus (r1, r2) -> eval_int r1 + eval_int r2
+      | Minus (r1, r2) -> eval_int r1 - eval_int r2
+      | Equal (r1, r2) -> raise TypeMismatch
+      | And (r1, r2) -> raise TypeMismatch
+      | Or (r1, r2) -> raise TypeMismatch
+      | Not r -> raise TypeMismatch)
+
+let rec eval_bool (r : result_value) : bool =
+  match r with
+  | FunctionResult _ | IntResult _ -> raise TypeMismatch
+  | BoolResult b -> b
+  | OpResult op_r -> (
+      match op_r with
+      | Plus (r1, r2) -> raise TypeMismatch
+      | Minus (r1, r2) -> raise TypeMismatch
+      | Equal (r1, r2) -> eval_int r1 = eval_int r2
+      | And (r1, r2) -> eval_bool r1 && eval_bool r2
+      | Or (r1, r2) -> eval_bool r1 || eval_bool r2
+      | Not r -> not (eval_bool r))
+
 let memo_cache = Hashtbl.create 1000
 
 (* can't do memoization like this in OCaml/Haskell; better laziness  *)
 (* laziness + memoization *)
-let rec eval_helper ?(is_lazy = false) le sigma =
-  match Hashtbl.find_opt memo_cache (le, sigma) with
+let rec eval_helper (e : expr) (sigma : int list) : result_value =
+  match Hashtbl.find_opt memo_cache (e, sigma) with
   | Some res -> res
   | None ->
-      let e, label = le in
       let eval_res =
         match e with
         (* Value *)
-        | Function (_, _) -> (le, sigma)
-        | Int _ -> (le, None)
-        | Bool _ -> (le, None)
+        | Value v -> (
+            match v with
+            | Function (ident, le', l) -> FunctionResult { f = v; l; sigma }
+            | Int i -> IntResult i
+            | Bool b -> BoolResult b)
         (* Application *)
-        | Appl (e1, _) -> (
-            if is_lazy then (le, None)
-            else
-              let fun_le = eval_helper e1 sigma |> fst |> fst in
-              match fun_le with
-              | Function (_, e') ->
-                  eval_helper e' (Some (label :: Option.get sigma)) ~is_lazy
-              | _ -> raise Unreachable [@coverage off])
-        | Var (Ident x) -> (
-            match get_outer_scope label |> fst with
-            | Function (Ident x', _) -> (
-                let sigma = Option.get sigma in
-                if x = x' then
-                  (* Var Local *)
-                  match get_expr (List.hd sigma) with
-                  | Appl (_, e2), _ ->
-                      eval_helper e2 (Some (List.tl sigma)) ~is_lazy
-                  | _ -> raise Unreachable [@coverage off]
-                else
-                  (* Var Non-Local *)
-                  match get_expr (List.hd sigma) with
-                  | Appl (e1, _), _ ->
-                      let (_, fun_label), sigma' =
-                        eval_helper e1 (Some (List.tl sigma))
-                      in
-                      eval_helper (Var (Ident x), fun_label) sigma' ~is_lazy
-                  | _ -> raise Unreachable [@coverage off])
+        | Appl (e1, _, app_l) -> (
```

Comment:
Gets evaluated during variable substitution because `eval_helper` returns a result value instead of an expression.